util/stopper.go

```diff
@@ -17,39 +17,96 @@
 
 package util
 
-import "sync"
+import (
+	"sync"
+	"sync/atomic"
+)
 
-// A Stopper provides a channel-based mechanism to stop a running
-// goroutine. Stopping occurs in two phases: the first is the request
-// to stop. The second is the confirmation by the goroutine that it
-// has stopped. Multiple goroutines can be stopped using the same
-// Stopper instance.
+// Closer is an interface for objects to attach to the stopper to
+// be closed once the stopper completes.
+type Closer interface {
+	Close()
+}
+
+// A Stopper provides a channel-based mechanism to stop an arbitrary
+// array of workers. Each worker is registered with the stopper via
+// the AddWorker() method. The system further tracks each task which
+// is outstanding by calling StartTask() when a task is started and
+// FinishTask() when completed.
+//
+// Stopping occurs in two phases: the first is the request to stop,
+// which moves the stopper into a draining phase. While draining,
+// calls to StartTask() return false, meaning the system is draining
+// and new tasks should not be accepted. When all outstanding tasks
+// have been completed via calls to FinishTask(), the stopper closes
+// its stopper channel, which signals all live workers that it's safe
+// to shut down. Once shutdown, each worker invokes SetStopped(). When
+// all workers have shutdown, the stopper is complete.
 type Stopper struct {
-	stopper chan struct{}
-	wg      sync.WaitGroup
+	stopper  chan struct{}
+	draining int32
+	drain    sync.WaitGroup
+	stop     sync.WaitGroup
+	mu       sync.Mutex // Protects the slice of Closers
+	closers  []Closer
 }
 
 // NewStopper returns an instance of Stopper. Count specifies how
```

Comment:
The `count` argument is no longer present so you can remove the second sentence.